### PR TITLE
feat: added special address with block number and block hash

### PR DESCRIPTION
**Pull request by Mohiiit**
Small fix for SNOS to run, it requires special address in the state diff
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Feature

## What is the current behavior?

Currently we were not adding the block hash and block number of previous block with special address which was breaking the SNOS

Resolves: #NA

## What is the new behavior?

- The above mention issue is resolved

## Does this introduce a breaking change?

No

### CODE REVIEW

File: crates/client/mempool/src/block_production.rs [hunk 1]
```diff
@@ -405,9 +405,32 @@ impl<Mempool: MempoolProvider> BlockProductionTask<Mempool> {
...
+                        address,
+                        storage_entries: vec![StorageEntry {
+                            key: Felt::from(prev_block_number),
+                            value: prev_block_hash,
+                        }],
+                    });
+                }
+                Ok(None) => {
+                    // this shouldn't happen, ideally should panic
+                    log::error!("No block hash found for block number {}", prev_block_number);
```
Comment (apoorvsadana) [inline crates/client/mempool/src/block_production.rs]: let's panic ig?

File: crates/client/mempool/src/block_production.rs [hunk 2]
```diff
@@ -405,9 +405,32 @@ impl<Mempool: MempoolProvider> BlockProductionTask<Mempool> {
...
         // Complete the block with full bouncer capacity.
         let start_time = Instant::now();
-        let (new_state_diff, _n_executed) =
+        let (mut new_state_diff, _n_executed) =
             self.continue_block(self.backend.chain_config().bouncer_config.block_max_capacity)?;
 
+        if block_n >= 10 {
+            let prev_block_number = block_n - 10;
+            match self.backend.get_block_hash(&BlockId::Number(prev_block_number)) {
+                Ok(Some(prev_block_hash)) => {
```
Comment (apoorvsadana) [inline crates/client/mempool/src/block_production.rs]: let's add a comment

File: crates/primitives/chain_config/src/chain_config.rs [hunk 3]
```diff
@@ -263,6 +263,8 @@ impl ChainConfig {
             chain_name: "Madara".into(),
             chain_id: ChainId::Other("MADARA_DEVNET".into()),
             sequencer_address: Felt::from_hex_unchecked("0x123").try_into().unwrap(),
+            block_time: Duration::from_secs(4),
+            pending_block_update_time: Duration::from_secs(1),
```
Comment (apoorvsadana) [inline crates/primitives/chain_config/src/chain_config.rs]:
```suggestion

```

**Review by apoorvsadana — APPROVED** [review]

File: crates/client/mempool/src/block_production.rs (hunk 2 above)
Comment (Mohiiit) [reply crates/client/mempool/src/block_production.rs]: added

File: crates/client/mempool/src/block_production.rs (hunk 1 above)
Comment (Mohiiit) [reply crates/client/mempool/src/block_production.rs]: resolved

File: crates/primitives/chain_config/src/chain_config.rs (hunk 3 above)
Comment (Mohiiit) [reply crates/primitives/chain_config/src/chain_config.rs]: resolved

**Review by apoorvsadana — APPROVED** [review]

**Comment by jbcaron** [comment]
This logic for block hashes is already implemented for blockifier with the StateAdapter, it should be changed accordingly while keeping a compatibility of re-execution of the mainnet.
https://github.com/madara-alliance/madara/blob/fe0993210671a21c3f19dbc654d8e5aab0f37c90/crates/client/exec/src/blockifier_state_adapter.rs#L33-L55

**Comment by apoorvsadana** [comment]
@jbcaron, I am mostly missing something but do we need this custom logic in the reader? The feeder gateway does return state diffs where it says `0x1` has the following storage keys and values. In full node mode, if we are applying these state diffs in the DB, won't normal fetching from DB work same as this custom logic we're using here?

**Comment by jbcaron** [comment]
we should just delete this code in the same PR but be sure that we can re-execute the mainnet blocks (I don't know from when this feature was introduced)
I checked on sepolia, there is no issue

**Comment by apoorvsadana** [comment]
> we should just delete this code in the same PR but be sure that we can re-execute the mainnet blocks 

makes sense, we can delete and see if the test cases pass + that we can resync a few hundred blocks on mainnet + get_storage_at on 0x1 address works as expected

EDIT: @Mohiiit has also already informed the visoft team for e2e cases for this specific case

**Review by Trantorian1 — DISMISSED** [review]

File: crates/client/mempool/src/block_production.rs [hunk 4]
```diff
@@ -405,9 +405,34 @@ impl<Mempool: MempoolProvider> BlockProductionTask<Mempool> {
...
         let start_time = Instant::now();
-        let (new_state_diff, _n_executed) =
+        let (mut new_state_diff, _n_executed) =
             self.continue_block(self.backend.chain_config().bouncer_config.block_max_capacity)?;
 
+        // SNOS requirement: For blocks >= 10, the hash of the block 10 blocks prior
+        // at address 0x1 with the block number as the key
+        if block_n >= 10 {
+            let prev_block_number = block_n - 10;
+            match self.backend.get_block_hash(&BlockId::Number(prev_block_number)) {
```
Comment (cchudant) [inline crates/client/mempool/src/block_production.rs]: this match can be simplified to
```rs
let prev_block_hash = self.backend.get_block_hash(&BlockId::Number(prev_block_number))
  .map_err(|err| Error::Unexpected(
      format!("Error fetching block hash for block {prev_block_number}: {err:#}").into()
  ))?
  .ok_or_else(|| Error::Unexpected(
      format!("No block hash found for block number {prev_block_number}").into()
  ))?;
let address = Felt::ONE;
new_state_diff.storage_diffs.push(ContractStorageDiffItem {
    address,
    storage_entries: vec![StorageEntry { key: Felt::from(block_n), value: prev_block_hash }],
});
  ```

Comment (Mohiiit) [reply crates/client/mempool/src/block_production.rs]: resolved

File: crates/client/rpc/src/versions/v0_7_1/methods/read/get_storage_at.rs [hunk 5]
```diff
@@ -45,17 +45,25 @@ pub fn get_storage_at(
...
+        block_number,
+        Some(DbBlockId::Number(num)) if num >= 10 && contract_address == Felt::ONE
+    );
+
+    if !skip_contract_check {
+        starknet
+            .backend
+            .get_contract_class_hash_at(&block_id, &contract_address)
+            .or_internal_server_error("Failed to check if contract is deployed")?
+            .ok_or(StarknetRpcApiError::ContractNotFound)?;
```
Comment (jbcaron) [inline crates/client/rpc/src/versions/v0_7_1/methods/read/get_storage_at.rs:60]: I think we should only return 0x0 in this case
we can therefore remove this check
this is already the implementation in the state adapter but maybe this logic should be in the db
@cchudant  what do you think?

Comment (Mohiiit) [reply crates/client/rpc/src/versions/v0_7_1/methods/read/get_storage_at.rs:60]: I checked with the juno rpc call and it has this behaviour hence implemented this

Comment (cchudant) [reply crates/client/rpc/src/versions/v0_7_1/methods/read/get_storage_at.rs:60]: apparently you have to return ContractNotFound
https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L216

but we need to bring this up into full node collab channel - getNonce also returns ContractNotFound but I think it's possible to increment the nonce of an undeclared contract

**Review by jbcaron — APPROVED** [review]

**Pull request merged**